book/src/proposals/booting-from-snapshot.md

```diff
@@ -0,0 +1,267 @@
+   validator stake has voted on at least once. Call this trusted bank `T`.
+   
+   3) The current working bank has slot number `S_current` > `S_n`
+
+8) Start the normal voting process, voting for any slot that satisfies
+`S_current` > `S_n`
+
+
+## Trusted Validator Set
+
+The booting validator needs to have a set of trusted validators that it uses to
+validate the snapshot and that it is caught up.  To guarantee that the booting
+validator connects to the cluster, the set needs to contain some threshold of
+validators that are still participating in consensus. This excludes validators
+that are running but are experiencing some attack (such as an eclipse attack)
+that prevents them from receiving slots, voting, or broadcasting their votes.
+
+### Verifying the Cluster is Building from the Snapshot
+
+The booting validator can confirm that the snapshot is valid and that the
+cluster it intends to join is building off of `S` by verifying the signatures
+of votes for `S`. The validator can calculate the commitment of its trusted
+validators, and at some threshold, such as 2/3 of trusted validators are locked
+out at least 2^31 slots, accept the snapshot.
+
+**Proof:** For the snapshot to be valid, the validator that made the snapshot
+must have rooted `S`, meaning that it must have observed at least 2/3 of stake
+having >=2^31 lockout on `S`. Any validators that are participating in
+consensus will eventually also observe the same lockout on `S`, and will
+therefore bring themselves up to >=2^31 lockout on `S`. Thus with any set of
+trusted validators meeting the conditions above, the booting validator will
+eventually observe the threshold commitment and accept the snapshot.
+
+### Verifying Booting Validator is Caught Up
+
+The booting validator should send canary transactions as described in the `Boot
+Procedure`. When the booting validator observes a canary transaction in a bank,
+and it observes at least one vote for that bank from a threshold (ie 2/3) of
+trusted validators, it can conclude that it has caught up to the trusted
+validators.
+
+
+## Determining `S_n` From a Snapshot and Tower
+
+Define the "safety" condition to be: 
+
+Given a `BankForks`, a validator can run some procedure to determine the first
+slot `S_n` that it can vote on without violating the lockouts of any vote in
+`L`. The procedure for this is:
+
+1) Find the earliest vote `V` in the tower for which `is_ancestor(V, S)` is not
+true.
+
+2) `V` and every vote after it needs to expire before the validator can vote on
+any descendant of `S`, so with all the votes `V_i` in tower where `V_i >= V`,
+`S_n = max(V_i + lockout(V_i))`. If no `V` exists, then `S_n = S`.
+
+
+## Achieving Safety
+
+Define the "Safety Criteria" to be:
+
+>If the validator has:
+>
+>  1. a `BankForks`
+>  2. any descendant of `S`, `S_d` that is present in `BankForks.banks`
+>  3. any slot `L_i` in tower
+>
+>the validator is able to determine `is_ancestor(L_i, S_d)`.
```

Comment:
This looks very nice than [before](https://github.com/solana-labs/solana/pull/6936#discussion_r354473504)!

Now with clearer understanding, this is my two cents! (Beware I'm not trained for the mathematical proof writing styles):

```
Define the "Safety Criteria" to be:

> If the validator has the following three things:
>
>  1. a `BankForks`
>  2. any descendant of `S`, `S_d` that is present in `BankForks.banks`
>  3. any slot `L_i` in tower
>
> then, the validator is able to determine `is_ancestor(L_i, S_d)`.
```

Or

```
Define the "Safety Criteria" to be:

The validator is able to determine `is_ancestor(L_i, S_d)` where `L_i` is any slot
in tower, `S_d` is any descendant of `S` and is present in `BankForks.banks`, and
in turn, a `BankForks` is present in the validator. 
```

Reply:
What do you think of this edit to your second option?

Reply:
Excellent! :heart: It now reads naturally!